### PR TITLE
[Under Powercut Testing] os/fs/smartfs : Add buffering of data to be written to smartfs file

### PR DESCRIPTION
**Pull request by amogh-samsung**
[Improve SmartFs Overwrite Method] These changes reduce the overhead in overwriting consecutive chunks of data of a file.
The case of appending data to the file is not affected as it is handled by a separate part of code, the modified code only deals with overwriting to the files.

- Hold the buffer from being committed to flash if the buffer is not completely full.
- The buffer will get synced when the file is closed or if another write request is
  issued which fills up the left over space in the buffer.

Signed-off-by: Taejun Kwon <tj80.kwon@samsung.com>, Amogh Hassija <a.hassija@samsung.com>

### CODE REVIEW

**Comment by amogh-samsung** [comment]
Kindly ignore smartfs_powercut_main.c file
Only uploaded for reference purposes
It will not be included in the merged commits.

File: apps/examples/smartfs_powercut/smartfs_powercut_main.c [hunk 1]
```diff
@@ -646,14 +646,139 @@ static int do_test(char *src, char *backup, char *backupdir, int bufsize, file_t
...
 }
-
+#endif
 #ifdef CONFIG_BUILD_KERNEL
 int main(int argc, FAR char *argv[])
 #else
 int smartfs_powercut_main(int argc, char *argv[])
 #endif
 {
+	//int ret;
```
Comment (deepaksrma) [inline apps/examples/smartfs_powercut/smartfs_powercut_main.c:656]: remove the comment

File: apps/examples/smartfs_powercut/smartfs_powercut_main.c [hunk 2]
```diff
@@ -646,14 +646,139 @@ static int do_test(char *src, char *backup, char *backupdir, int bufsize, file_t
...
+		lseek(fd, 0, SEEK_SET);
+		for (int i = 0; i < j/2; i++) {
+			write(fd, buf, 1024);
+		}
+		printf("Write complete\n");
+		close(fd);
+		printf("File closed\n");
+
+		fd = open("/mnt/test", O_RDOK);
+                if (fd < 0)
```
Comment (deepaksrma) [inline apps/examples/smartfs_powercut/smartfs_powercut_main.c:683]: add brackets {}

File: apps/examples/smartfs_powercut/smartfs_powercut_main.c [hunk 3]
```diff
@@ -646,14 +646,139 @@ static int do_test(char *src, char *backup, char *backupdir, int bufsize, file_t
...
+                        for (int k = 0; k < 1024; k++) {
+                                printf("%c", tmp[k]);
+                                if ((k+1)%100 == 0)
+                                        printf("\n");
+                        }
+                        printf("\n");
+                }
+		close(fd);
+
+		//unlink("/mnt/test");
```
Comment (deepaksrma) [inline apps/examples/smartfs_powercut/smartfs_powercut_main.c:778]: remove commented code and check formatting

**Comment by amogh-samsung** [comment]
@deepaksrma please do not review powercut_main.c file as mentioned above
It will not be a part of the final commit

File: apps/examples/smartfs_powercut/smartfs_powercut_main.c [hunk 4]
```diff
@@ -646,14 +646,139 @@ static int do_test(char *src, char *backup, char *backupdir, int bufsize, file_t
...
 #endif
 {
+	//int ret;
+	int fd;
 	int ret;
+	char buf[1024];
+	char tmp[1024];
+
+
+	for (int j = 10; j <= 10; j++)  {
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply apps/examples/smartfs_powercut/smartfs_powercut_main.c:663]: What is the meaning of the loop which is from 10 to 10? Why do you use for loop?

File: apps/examples/smartfs_powercut/smartfs_powercut_main.c [hunk 5]
```diff
@@ -646,14 +646,139 @@ static int do_test(char *src, char *backup, char *backupdir, int bufsize, file_t
...
+		for (int i = 0; i < j; i++) {
+			write(fd, buf, 1024);
+		}
+		close(fd);
+
+		memset(buf, '0', 1024);
+		printf("##### Start Partial Overwrite Case #####\n");
+		fd = open("/mnt/test", O_WROK);
+		lseek(fd, 0, SEEK_SET);
+		for (int i = 0; i < j/2; i++) {
```
Comment (jeongchanKim) [inline apps/examples/smartfs_powercut/smartfs_powercut_main.c:675]: Please add space before and after `/`.

File: apps/examples/smartfs_powercut/smartfs_powercut_main.c [hunk 6]
```diff
@@ -646,14 +646,139 @@ static int do_test(char *src, char *backup, char *backupdir, int bufsize, file_t
...
+		}
+		printf("Write complete\n");
+		close(fd);
+		printf("File closed\n");
+
+		fd = open("/mnt/test", O_RDOK);
+                if (fd < 0)
+                        printf("Unable to open file\n");
+                for (int i = 0; i < j; i++) {
+                        ret = read(fd, tmp, 1024);
```
Comment (jeongchanKim) [inline apps/examples/smartfs_powercut/smartfs_powercut_main.c:686]: Is it okay to read the `fd` when it was failed to open?

File: apps/examples/smartfs_powercut/smartfs_powercut_main.c [hunk 7]
```diff
@@ -646,14 +646,139 @@ static int do_test(char *src, char *backup, char *backupdir, int bufsize, file_t
...
+                        if (ret < 1024)
+                                printf("Could not read 1024 bytes in itr %d\n", i);
+                        for (int k = 0; k < 1024; k++) {
+                                printf("%c", tmp[k]);
+                                if ((k+1)%100 == 0)
```
Comment (jeongchanKim) [inline apps/examples/smartfs_powercut/smartfs_powercut_main.c:691]: Please follow the coding rule about space.

File: apps/examples/smartfs_powercut/smartfs_powercut_main.c [hunk 8]
```diff
@@ -646,14 +646,139 @@ static int do_test(char *src, char *backup, char *backupdir, int bufsize, file_t
...
+                }
+                printf("Write complete\n");
+                close(fd);
+		printf("File closed\n");
+
+		fd = open("/mnt/test", O_RDOK);
+		if (fd < 0)
+			printf("Unable to open file\n");
+		for (int i = 0; i < j; i++) {
+			ret = read(fd, tmp, 1024);
```
Comment (jeongchanKim) [inline apps/examples/smartfs_powercut/smartfs_powercut_main.c:713]: Same as above. Please check the `fd` is valid or not.

**Review by jeongchanKim — COMMENTED** [review]
Please split the commit, functionality changes and apps.

**Comment by amogh-samsung** [comment]
> Please split the commit, functionality changes and apps.

Please check my comment, powercut_main.c will not be a part of the final commit
It is for current reference only.

File: apps/examples/smartfs_powercut/smartfs_powercut_main.c [hunk 9]
```diff
@@ -646,14 +646,139 @@ static int do_test(char *src, char *backup, char *backupdir, int bufsize, file_t
...
+                                printf("%c", tmp[k]);
+                                if ((k+1)%100 == 0)
+                                        printf("\n");
+                        }
+                        printf("\n");
+                }
+		close(fd);
+
+		memset(buf, '3', 1024);
+                printf("##### Start Overwrite-Append Case #####\n");
```
Comment (dineshdas-samsung) [inline apps/examples/smartfs_powercut/smartfs_powercut_main.c:752]: Please maintain the indentation !

File: os/fs/smartfs/smartfs_smart.c [hunk 10]
```diff
@@ -651,6 +653,10 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer, size_t
 
 	header = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
 	byteswritten = 0;
+
+	/* As long as the current file pointer positions for writing is less than the total data length in the file, we are doing and overwrite operation
+	 * When the file pointer positins surpasses the file's data length, it becomes an append operation which will be handled by a separate function
```
Comment (dineshdas-samsung) [inline os/fs/smartfs/smartfs_smart.c:658]: /* As long as the current file pointer "position" for writing is less than the total data length in the file, we are doing and "an"overwrite operation
	 * When the file pointer "position" surpasses the file's data length, it becomes an append operation which will be handled by a separate function

**Comment by Taejun-Kwon** [comment]
Please Add below to pr title
'to improve overwrite'

**Comment by davidfather** [comment]
> Please Add below to pr title
> 'to improve overwrite'

@amogh-samsung Could you describe why this commit does not affect write performance but improve overwrite performance?

**Comment by Taejun-Kwon** [comment]
> > Please Add below to pr title
> > 'to improve overwrite'
> 
> @amogh-samsung Could you describe why this commit does not affect write performance but improve overwrite performance?

please see while loop in line 640
```
while ((sf->filepos < sf->entry.datalen) && (buflen > 0)) {
```
This means that current position of file is not greater or equal than length of file, so it just affect only to overwrite case.

File: os/fs/smartfs/smartfs_smart.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (amogh-samsung) [reply os/fs/smartfs/smartfs_smart.c]: Done

File: os/fs/smartfs/smartfs_smart.c [hunk 11]
```diff
@@ -637,58 +638,60 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer, size_t
...
 
-	/* First test if we are overwriting an existing location or writing to
-	 * a new one. */
-
 	header = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
 	byteswritten = 0;
+
+	/* First, check if we are overwriting the existing file data or appending more
+	 * As long as the current file pointer position for writing is less than
+	 * the total data length in the file, we are doing and overwrite operation.
```
Comment (abhishek-samsung) [inline os/fs/smartfs/smartfs_smart.c:653]: Please change "doing and overwrite operation" --> "doing an overwrite operation"

File: os/fs/smartfs/smartfs_smart.c [hunk 12]
```diff
@@ -637,58 +639,62 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer, size_t
...
 		ret = -EACCES;
 		goto errout_with_semaphore;
 	}
 
-	/* First test if we are overwriting an existing location or writing to
-	 * a new one. */
-
 	header = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
 	byteswritten = 0;
+	size = sizeof(struct smartfs_chain_header_s);
```
Comment (Taejun-Kwon) [inline os/fs/smartfs/smartfs_smart.c:651]: it never changed, you'd better initialize it directly in line 612.

Comment (amogh-samsung) [reply os/fs/smartfs/smartfs_smart.c:651]: Ohkay

**Review by Taejun-Kwon — APPROVED** [review]

**Pull request merged**